prow/servemux/servemux.go

```diff
@@ -0,0 +1,399 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package servemux
+
+import (
+	"bytes"
+	"encoding/json"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/sirupsen/logrus"
+
+	"k8s.io/test-infra/prow/config"
+	"k8s.io/test-infra/prow/github"
+	"k8s.io/test-infra/prow/interrupts"
+	"k8s.io/test-infra/prow/pluginhelp"
+	pluginhelp_externalplugins "k8s.io/test-infra/prow/pluginhelp/externalplugins"
+	pluginhelp_hook "k8s.io/test-infra/prow/pluginhelp/hook"
+
+	"k8s.io/test-infra/prow/plugins"
+)
+
+const (
+	eventTypeField = "event-type"
+)
+
+type ServeMuxHandlerOptions struct {
+	wg *sync.WaitGroup
+
+	serveMuxHandler *serveMuxHandler
+
+	endpoint     string
+	port         int
+	httpServeMux *http.ServeMux
+}
+
+func (o *ServeMuxHandlerOptions) Validate() error {
+	if !strings.HasPrefix(o.endpoint, "/") {
+		return fmt.Errorf("endpoint %s is not a valid url path", o.endpoint)
+	}
+	return nil
+}
+
+func (o *ServeMuxHandlerOptions) Bind(fs *flag.FlagSet) {
+	fs.StringVar(&o.endpoint, "endpoint", "/hook", "The endpoint path where the http server will listen to")
+	fs.IntVar(&o.port, "port", 8888, "Port to listen on.")
+}
+
+func (o *ServeMuxHandlerOptions) Initialize(hmacTokenGenerator func() []byte, log *logrus.Entry) {
+	var wg sync.WaitGroup
+
+	o.wg = &wg
+	o.httpServeMux = http.NewServeMux()
+	o.serveMuxHandler = &serveMuxHandler{
+		hmacTokenGenerator: hmacTokenGenerator,
+		log:                log,
+		metrics:            NewMetrics(),
```

Comment:
Why is `NewMetrics` exported if it can not be set?

Reply:
other components can use these metrics in their handle event funcitons